ccsdspy/converters.py

```diff
@@ -0,0 +1,189 @@
+"""This class hold the implementation of the converter sysytem, which applies 
+post-process to decoded packet fields. This post-processing includes applying
+linear/polynomial calibration curves, dictionary replacement, and time parsing.
+"""
+
+import numpy as np
+
+
+class EnumConverterMissingKey(RuntimeError):
+    """During conversion a value was encountered which did not have a
+    corresponding key in the replacement dictionary.
+    """
+
+
+class Converter:
+    """Base class for all converter objects.
+
+    This class is extended to create converters, and users may extend this
+    class to write their own custom converters.
+
+    To write a converter, one must create a subclass and override either the method
+    `convert_many(field_array)` or `convert_one(field_array)`. The
+    `convert_many(field_array)` method implements the conversion for an entire
+    sequence of decoded packet field values in a single call (this is sometimes
+    faster), while the `convert_one(field_array)` method implements the
+    conversion for a single packet field at a time (this is sometimes easier to
+    write).
+    """
+
+    def __init__(self):
+        raise NotImplementedError("This is a base class not meant to be instantiated directly")
+
+    def convert_many(self, field_array):
+        """Convert a sequence of decoded packet field values.
+
+        This default implementation loops over the elements and defers the
+        conversion to `self.convert_one(field_array[i])`.
+
+        Parameters
+        ----------
+        field_array : NumPy array
+            decoded packet field values, must have at least one dimension
+
+        Returns
+        -------
+        converted_field_array : NumPy array
+            converted form of the decoded packet field values
+        """
+        converted_field_array = []
+
+        for i in range(field_array.shape[0]):
+            converted = self.convert_one(field_array[i])
+            converted_field_array.append(converted)
+
+        # Let NumPy infer the type from the elements.
+        converted_field_array = np.array(converted_field_array)
+
+        return converted_field_array
+
+    def convert_one(self, field_value):
```

Comment:
I'm not sure about providing two different ways to convert. Since most everything can be converted into arrays why have ```convert_one``` at all? Seems to add unnecessary complexity.

Reply:
You're right, I think our users are smart enough to write their own for loops. I'll take this out.